packages/eslint-plugin/src/rules/require-object-type-annotations.ts

```diff
@@ -0,0 +1,55 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import type { RuleListener } from '@typescript-eslint/utils/src/ts-eslint';
+import type * as ts from 'typescript';
+
+import * as util from '../util';
+
+export default util.createRule({
+  name: 'require-object-type-annotations',
+  defaultOptions: [],
+  meta: {
+    docs: {
+      description: '',
+      requiresTypeChecking: true,
+      recommended: false,
+    },
+    messages: {
+      forbidden: 'Object is missing type annotation.',
+    },
+    schema: [],
+    type: 'problem',
+  },
+  create: (context): RuleListener => {
+    const listener = (esNode: TSESTree.ObjectExpression): void => {
+      const parserServices = util.getParserServices(context);
+      const checker = parserServices.program.getTypeChecker();
+
+      const tsNode: ts.ObjectLiteralExpression =
+        parserServices.esTreeNodeToTSNodeMap.get(esNode);
+
+      const type = checker.getTypeAtLocation(tsNode);
+
+      // Allow empty objects
+      if (type.getProperties().length === 0) {
```

Comment:
Note that most types will have properties!
I believe that only the following types will not have any properties:
- `null`
- `undefined`
- `object`
- `{}`

It would probably be a good idea to check the `type.flags` for things like `ts.TypeFlags.StringLike`, `ts.TypeFlags.NumberLike`, etc to help weed out things we definitely want to ignore.

Reply:
Is it possible to have an `ObjectExpression` node where the type is `StringLike` or `NumberLike`? 🤔

For context, I was only intending to catch `{}` with this branch.